File: app/src/main/java/pe/andy/bookholic/adapter/LibraryAdapter.java
package pe.andy.bookholic.adapter;

import android.content.Context;
import android.graphics.Color;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.core.content.ContextCompat;
import androidx.recyclerview.widget.RecyclerView;

import java.util.List;
import java.util.Locale;

import lombok.Getter;
import lombok.Setter;
import pe.andy.bookholic.R;
import pe.andy.bookholic.databinding.LibraryItemBinding;
import pe.andy.bookholic.searcher.LibrarySearchTask;

import static android.view.View.GONE;

public class LibraryAdapter extends RecyclerView.Adapter<LibraryAdapter.LibraryViewHolder>  {

    private Context mContext;
    @Getter @Setter
    private List<LibrarySearchTask> tasks;

    private static int textDefault;
    private static int textMuted;

    public LibraryAdapter(Context mContext, List<LibrarySearchTask> tasks) {
        this.mContext = mContext;
        this.tasks = tasks;

        textMuted = ContextCompat.getColor(mContext, R.color.textMuted);
        textDefault = ContextCompat.getColor(mContext, android.R.color.tab_indicator_text);
    }

    @NonNull
    @Override
    public LibraryAdapter.LibraryViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        LibraryItemBinding binding = LibraryItemBinding.inflate(
                LayoutInflater.from(mContext), parent, false);

        return new LibraryViewHolder(binding.getRoot(), binding);
    }


    @Override
    public void onBindViewHolder(@NonNull LibraryAdapter.LibraryViewHolder holder, int position) {
        LibrarySearchTask task = this.tasks.get(position);

        holder.binding.libraryName.setText(task.getLibraryName());
        holder.binding.libraryIcon.setVisibility(GONE);
        holder.binding.searchProgress.setVisibility(GONE);
        holder.binding.searchDone.setVisibility(GONE);
        holder.binding.searchFail.setVisibility(GONE);

        View icon = holder.binding.libraryIcon;
        switch(task.getSearchStatus()) {
            case PROGRESS:
                icon = holder.binding.searchProgress;
                break;
            case DONE:
                if (task.getResultCount() > 0) {
                    icon = holder.binding.searchDone;
                }
                break;
            case FAIL:
                icon = holder.binding.searchFail;
        }
        icon.setVisibility(View.VISIBLE);

        // UI를 초기화
        holder.binding.libraryName.setTextColor(textDefault);
        holder.binding.searchCount.setVisibility(GONE);

        if (task.isSearchDone()) {
            if (task.getResultCount() == 0) {
                holder.binding.libraryName.setTextColor(textMuted);
            }
            else {
                holder.binding.searchCount.setText(String.format(Locale.KOREA, "%d", task.getResultCount()));
                holder.binding.searchCount.setTextColor(Color.WHITE);
                holder.binding.searchCount.setVisibility(View.VISIBLE);
            }
        }
    }

    @Override
    public int getItemCount() {
        return (this.tasks == null) ? 0 : this.tasks.size();
    }

    class LibraryViewHolder extends RecyclerView.ViewHolder {
        LibraryItemBinding binding;

        public LibraryViewHolder(@NonNull View itemView, LibraryItemBinding binding) {
            super(itemView);
            this.binding = binding;
        }
    }

}

File: app/src/main/java/pe/andy/bookholic/adapter/LibraryAdapter.kt
package pe.andy.bookholic.adapter

import android.graphics.Color
import android.view.LayoutInflater
import android.view.View.GONE
import android.view.View.VISIBLE
import android.view.ViewGroup
import androidx.core.content.ContextCompat
import androidx.recyclerview.widget.RecyclerView
import pe.andy.bookholic.R
import pe.andy.bookholic.databinding.LibraryItemBinding
import pe.andy.bookholic.searcher.LibrarySearchTask
import pe.andy.bookholic.searcher.LibrarySearchTask.LibrarySearchStatus.DONE
import pe.andy.bookholic.searcher.LibrarySearchTask.LibrarySearchStatus.FAIL
import pe.andy.bookholic.searcher.LibrarySearchTask.LibrarySearchStatus.PROGRESS

class LibraryAdapter(var libraries: MutableList<LibrarySearchTask>) : RecyclerView.Adapter<LibraryAdapter.LibraryViewHolder>() {

    var textMuted: Int = 0
    var textDefault: Int = 0

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): LibraryViewHolder {
        val binding = LibraryItemBinding.inflate(
                LayoutInflater.from(parent.context), parent, false)

        textMuted = ContextCompat.getColor(parent.context, R.color.textMuted)
        textDefault = ContextCompat.getColor(parent.context, android.R.color.tab_indicator_text)

        return LibraryViewHolder(binding)
    }

    override fun getItemCount(): Int = libraries.size

    override fun onBindViewHolder(holder: LibraryViewHolder, position: Int) {
        val library = libraries[position]
        holder.bind(library)
    }

    fun set(list: List<LibrarySearchTask>) {
        libraries.clear()
        libraries.addAll(list)
        notifyDataSetChanged()
    }

    fun clear() {
        libraries.clear()
        notifyDataSetChanged()
    }

    fun refresh() {
        notifyDataSetChanged()
    }

    inner class LibraryViewHolder(val binding: LibraryItemBinding)
        : RecyclerView.ViewHolder(binding.root) {

        fun bind(library: LibrarySearchTask) {

            with(binding) {
                libraryName.text = library.library.name
                libraryIcon.visibility = GONE
                searchProgress.visibility = GONE
                searchDone.visibility = GONE
                searchFail.visibility = GONE

                when (library.searchStatus) {
                    PROGRESS -> searchProgress
                    DONE -> when {
                        library.resultCount > 0 -> searchDone
                        else -> libraryIcon
                    }
                    FAIL -> searchFail
                    else -> libraryIcon
                }.run {
                    visibility = VISIBLE
                }

                // UI를 초기화
                libraryName.setTextColor(textDefault)
                searchCount.visibility = GONE

                // 검색이 끝나면 검색결과에 따른 처리
                if (library.searchStatus == DONE) {
                    when(library.resultCount) {
                        0 -> libraryName.setTextColor(textMuted)
                        else -> {
                            with(searchCount) {
                                text = library.resultCount.toString()
                                setTextColor(Color.WHITE)
                                visibility = VISIBLE
                            }
                        }
                    }
                }
            }
        }
    }
}